### PR TITLE
Blur algorithm overhaul

### PR DESCRIPTION
**Pull request by MrSmile**
My initial blur algorithm was overcomplicated due to (mostly arbitrary) decision to use filters not larger than 9-tap. That leads to use of an additional stage (prefilter) and irregular main filter pattern (1235, 1246). Over time I realized that it's not worth it and did some experiments. And now it's time to apply my new understanding to libass!

Also now I have AVX2-capable CPU and more practice with AVX2 assembly so there are some improvements in that area. Overall this new algorithm should have slightly better precision and performance than old.

### CODE REVIEW

File: libass/ass.c [hunk 1]
```diff
@@ -19,20 +19,12 @@
...
-#include <errno.h>
-#include <sys/types.h>
-#include <sys/stat.h>
 #include <inttypes.h>
 
 #ifdef CONFIG_ICONV
 #include <iconv.h>
 #endif
 
-#include "ass.h"
```
Comment (rcombs) [inline libass/ass.c:35]: I'm very confused. This file is literally defining the functions declared in this header. How is it inappropriate for it to include it?

File: libass/ass.c [hunk 2]
```diff
@@ -19,20 +19,12 @@
 #include "config.h"
 #include "ass_compat.h"
 
-#include <stdio.h>
-#include <stdlib.h>
```
Comment (rcombs) [inline libass/ass.c:23]: This file uses a lot of stuff in these system headers; just because it happens to include something else that happens to also include these doesn't mean we shouldn't include them here. I don't like this commit apart from the removal of includes that are completely unused in a particular file (like FT_OUTLINE_H in files that don't use anything from that header).

File: libass/ass_blur.c [hunk 3]
```diff
@@ -26,16 +26,15 @@
...
  * Experiment shows that downscaling of factor 2 with kernel [1, 5, 10, 10, 5, 1] and
  * corresponding upscaling are enough for 8-bit precision.
  *
- * For central filter here is used generic 9-tap filter with one of 3 different patterns
- * combined with one of optional prefilters with fixed kernels. Kernel coefficients
- * of the main filter are obtained from solution of least squares problem
- * for Fourier transform of resulting kernel.
+ * For central filter here is used generic filter with 5 different kernel width (9 to 17-tap).
+ * Kernel coefficients of that filter are obtained from the solution of least squares problem
+ * for Fourier transform of the resulting kernel.
```
Comment (rcombs) [inline libass/ass_blur.c:37]:
```suggestion
 * Here we use generic filters with 5 different kernel widths (9 to 17-tap).
 * Kernel coefficients of that filter are obtained from the solution of the least-squares problem
 * for the Fourier transform of the resulting kernel.
```

File: libass/ass_blur.c [hunk 4]
```diff
@@ -26,16 +26,15 @@
  * Cascade Blur Algorithm
  *
  * The main idea is simple: to approximate gaussian blur with large radius
- * you can downscale, then apply filter with small pattern, then upscale back.
+ * you can downscale, then apply filter with relativly small pattern, then upscale back.
```
Comment (rcombs) [inline libass/ass_blur.c:29]:
```suggestion
 * The main idea is simple: to approximate a gaussian blur with large radius,
 * you can scale down, apply a filter with a relatively small pattern, then scale back up.
```

File: libass/ass_blur.c [hunk 5]
```diff
@@ -26,16 +26,15 @@
...
  *
  * To achieve desired precision down/upscaling should be done with sufficiently smooth kernel.
  * Experiment shows that downscaling of factor 2 with kernel [1, 5, 10, 10, 5, 1] and
  * corresponding upscaling are enough for 8-bit precision.
```
Comment (rcombs) [inline libass/ass_blur.c:33]:
```suggestion
 * To achieve the desired precision, scaling should be done with sufficiently smooth kernel.
 * Experiments show that downscaling of factor 2 with kernel [1, 5, 10, 10, 5, 1] and
 * corresponding upscaling are enough for 8-bit precision.
```

File: libass/ass_blur.c [hunk 6]
```diff
@@ -75,10 +72,10 @@ inline static void copy_line(int16_t *buf, const int16_t *ptr, uintptr_t offs, u
 void ass_stripe_unpack_c(int16_t *dst, const uint8_t *src, ptrdiff_t src_stride,
                          uintptr_t width, uintptr_t height)
 {
-    for (uintptr_t y = 0; y < height; ++y) {
+    for (uintptr_t y = 0; y < height; y++) {
```
Comment (rcombs) [inline libass/ass_blur.c:75]: Would be nice to split these little stylistic changes into a separate commit.

File: libass/ass_blur.c [hunk 7]
```diff
@@ -59,9 +58,7 @@ inline static const int16_t *get_line(const int16_t *ptr, uintptr_t offs, uintpt
 
 inline static void copy_line(int16_t *buf, const int16_t *ptr, uintptr_t offs, uintptr_t size)
 {
-    ptr = get_line(ptr, offs, size);
-    for (int k = 0; k < STRIPE_WIDTH; ++k)
-        buf[k] = ptr[k];
+    memcpy(buf, get_line(ptr, offs, size), STRIPE_WIDTH * sizeof(buf[0]));
```
Comment (rcombs) [inline libass/ass_blur.c:61]: Similarly, this can be split out.

File: libass/ass_blur.c [hunk 8]
```diff
@@ -250,404 +247,153 @@ void ass_expand_vert_c(int16_t *dst, const int16_t *src,
...
 {
-    uintptr_t dst_width = src_width + 4;
+    uintptr_t dst_width = src_width + 2 * n;
     uintptr_t size = ((src_width + STRIPE_MASK) & ~STRIPE_MASK) * src_height;
     uintptr_t step = STRIPE_WIDTH * src_height;
 
     uintptr_t offs = 0;
-    int16_t buf[2 * STRIPE_WIDTH];
-    int16_t *ptr = buf + STRIPE_WIDTH;
+    int16_t buf[3 * STRIPE_WIDTH], *ptr = buf + 2 * STRIPE_WIDTH;
```
Comment (rcombs) [inline libass/ass_blur.c:277]: I think this'd be clearer if kept as 2 separate lines.

File: libass/ass_blur.c [hunk 9]
```diff
@@ -693,142 +429,95 @@ static void coeff_filter(double *coeff, int n, const double kernel[4])
...
-    double vec[4];
-    for (int i = 0; i < 4; ++i)
-        vec[i] = mat_freq[0] - mat_freq[index[i]] - vec_freq[0] + vec_freq[index[i]];
-
-    for (int i = 0; i < 4; ++i) {
+    double mat_freq[17] = { kernel[0], kernel[1], kernel[2], kernel[3] };
+    coeff_filter(mat_freq, 7, kernel);
+
+    double vec_freq[12];
+    calc_gauss(vec_freq, n + 4, r2 * mul);
```
Comment (rcombs) [inline libass/ass_blur.c:480]: Worth an `assert(n + 4 <= sizeof(vec_freq)/sizeof(vec_freq[0]))` here.

File: libass/ass_blur.c [hunk 10]
```diff
@@ -861,38 +547,28 @@ bool ass_gaussian_blur(const BitmapEngine *engine, Bitmap *bm, double r2)
...
         index ^= 1;
     }
-    if (blur.prefilter) {
-        engine->pre_blur_horz[blur.prefilter - 1](buf[index ^ 1], buf[index], w, h);
-        w += 2 * blur.prefilter;
-        index ^= 1;
-    }
-    engine->main_blur_horz[blur.filter](buf[index ^ 1], buf[index], w, h, blur.coeff);
-    w += 2 * blur.filter + 8;
+    engine->blur_horz[blur.radius - 4](buf[index ^ 1], buf[index], w, h, blur.coeff);
```
Comment (rcombs) [inline libass/ass_blur.c:560]: `assert(blur.radius >= 4);`

File: libass/ass.c (hunk 1 above)
Comment (MrSmile) [reply libass/ass.c:35]: It's included through `ass_utils.h`. I don't think that header list has any impact on code clarity, so it's better for compiler to touch hard drive less.

File: libass/ass.c (hunk 2 above)
Comment (MrSmile) [reply libass/ass.c:23]: `ass_utils.h` is essentially a sort of precompiled header with bunch of standard system includes that get included everywhere. So I think that it's not much sense to include them repeatedly. But I'm not holding that commit too dearly so if there is a consensus that excessive includes are necessary I'll drop it.

**Comment by MrSmile** [comment]
> How's the perf look in benchmarking?

Slightly better on my large blur benchmark. But it's a complex test with bunch of blur radii, ideally blur should be benchmarked individually for each radius and for different CPU/OSes too.

> Can we add a couple higher-value \blur and \be tests to the compare infrastructure and use those to validate this?

I have a [special test](https://github.com/MrSmile/libass-tests/tree/master/blur-test) for that. Should trigger most of blur codepaths.

File: libass/ass_blur.c (hunk 7 above)
Comment (MrSmile) [reply libass/ass_blur.c:61]: I think that quite falls under "Simplify blur algorithm". If I split that into another commit then assembly modifications of the old functions (shrink/expand) should be separated too.

File: libass/ass_blur.c (hunk 9 above)
Comment (MrSmile) [reply libass/ass_blur.c:480]: `n + 4` is already copy-paste so using `sizeof(vec_freq)/sizeof(vec_freq[0])` instead of simply `12` don't make much sense. I think I'll go with `assert(n <= 8)` at the start of the function.

File: libass/ass_blur.c (hunk 4 above)
Comment (MrSmile) [reply libass/ass_blur.c]: Done

File: libass/ass_blur.c (hunk 6 above)
Comment (MrSmile) [reply libass/ass_blur.c:75]: Done

File: libass/ass_blur.c (hunk 8 above)
Comment (MrSmile) [reply libass/ass_blur.c]: Done

File: libass/ass_blur.c (hunk 10 above)
Comment (MrSmile) [reply libass/ass_blur.c:564]: Done

**Comment by MrSmile** [comment]
I've rebased and moved the first unrelated commit to a separate PR #430.

**Review by rcombs — APPROVED** [review]

**Comment by astiob** [comment]
I’m just gonna trust you on this one.
